mu/virtual_environment.py

```diff
@@ -0,0 +1,544 @@
+        #
+        with open(self.BASELINE_PACKAGES_FILEPATH, encoding="utf-8") as f:
+            try:
+                return json.load(f)
+            except json.decoder.JSONDecodeError:
+                logger.exception("Unable to read baseline packages")
+                return []
+
+    def install_user_packages(self, packages, slots=Process.Slots()):
+        logger.info("Installing user packages: %s", ", ".join(packages))
+        self.pip.install(
+            packages,
+            slots=slots,
+            upgrade=True,
+        )
+
+    def remove_user_packages(self, packages, slots=Process.Slots()):
+        logger.info("Removing user packages: %s", ", ".join(packages))
+        self.pip.uninstall(
+            packages,
+            slots=slots,
+        )
+
+    def full_pythonpath(self):
+        """
+        Given an interpreter from a virtualenv, returns a PYTHONPATH containing
+        both the virtualenvs paths and then the paths from Mu's own sys.path.
+        """
+        #
+        # FIXME: not sure where this is used; maybe only Kernel creation?
+        #
+        result = self.run_python(
+            "-c", "import sys; print('\\n'.join(sys.path))"
+        )
+        #
+        # Using list rather than set to preserve seach order
+        #
+        paths = []
+        for p in [line.strip() for line in result.split("\n")]:
+            if p not in paths:
+                paths.append(p)
+        for p in sys.path:
+            if p not in paths:
+                paths.append(p)
+        return os.pathsep.join(paths)
+
+    def installed_packages(self):
+        """
+        List all the third party modules installed by the user in the venv
+        containing the referenced Python interpreter.
+        """
+        logger.info("Discovering installed third party modules in venv.")
+
+        #
+        # FIXME: Basically we need a way to distinguish between installed
+        # baseline packages and user-added packages. The baseline_packages
+        # in this class (or, later, from modes) are just the top-level classes:
+        # flask, pgzero etc. But they bring in many others further down. So:
+        # we either need to keep track of what's installed as part of the
+        # baseline install; or to keep track of what's installed by users.
+        # And then we have to hold those in the settings file
+        # The latter is probably easier.
+        #
+
+        baseline_packages = [
+            name for name, version in self.baseline_packages()
+        ]
+        user_packages = []
+        for package, version in self.pip.installed():
+            logger.info(package)
+            if package not in baseline_packages:
+                user_packages.append(package)
+
+        return baseline_packages, user_packages
+
+
+#
+# Create a singleton virtual environment to be used throughout
+# the application
+#
+venv = VirtualEnvironment(config.VENV_DIR)

```

Comment:
Just looked into Singleton pattern for python, and discovered this comment on Stackoverflow "This only works if all the imports happen the same way. `import project.module` and `import .module` will run the code twice." https://stackoverflow.com/questions/6760685/creating-a-singleton-in-python#comment80323547_6760726

Just something to be aware of, when using this rather simple singleton implementation.

Reply:
Good catch, @dybber. I was technically aware of this, but hadn't thought it through wrt the implementation. In fact, even if we do start up two instances of the venv, it's unlikely to be a problem as they'd both be pointing to the same underlying venv. And I'm not aware of a situation where they could run in parallel, possibly interfering (or seeing locking errors).

But thanks for picking it up; I'll do a sweep and make sure I'm importing the module in the same way across the board.

Reply:
I guess a REPL can be open, while a user installs a package? I'm not sure if that counts